contracts/protocol/libraries/logic/PoolExtendedLogic.sol

```diff
@@ -0,0 +1,442 @@
+                userConfig,
+                DataTypes.ExecuteBorrowParams({
+                    asset: weth,
+                    user: msg.sender,
+                    onBehalfOf: msg.sender,
+                    amount: totalBorrow,
+                    referralCode: 0,
+                    releaseUnderlying: false,
+                    reservesCount: ps._reservesCount,
+                    oracle: oracle,
+                    priceOracleSentinel: poolAddressProvider
+                        .getPriceOracleSentinel()
+                })
+            );
+        }
+
+        //transfer currency to keeper
+        if (totalBorrow > 0) {
+            DataTypes.TimeLockParams memory timeLockParams;
+            IPToken(ps._reserves[weth].xTokenAddress).transferUnderlyingTo(
+                address(this),
+                totalBorrow,
+                timeLockParams
+            );
+            IWETH(weth).withdraw(totalBorrow);
+        }
+        Helpers.safeTransferETH(keeper, msg.value + totalBorrow);
+    }
+
+    function initiateBlurExchangeRequest(
+        DataTypes.PoolStorage storage ps,
+        DataTypes.UserConfigurationMap storage userConfig,
+        DataTypes.BlurBuyWithCreditRequest calldata request,
+        uint256 remainingETH,
+        uint256 requestFeeRate,
+        address oracle
+    ) internal returns (uint256) {
+        bytes32 requestHash = _calculateBlurExchangeRequestHash(request);
+        uint256 requestFee = request.listingPrice.percentMul(requestFeeRate);
+        ValidationLogic.validateInitiateBlurExchangeRequest(
+            ps._reserves[request.collection],
+            request,
+            ps._blurExchangeRequestStatus[requestHash],
+            remainingETH,
+            requestFee,
+            oracle
+        );
+
+        //mint nToken to release credit value
+        DataTypes.ERC721SupplyParams[]
+            memory tokenData = new DataTypes.ERC721SupplyParams[](1);
+        tokenData[0] = DataTypes.ERC721SupplyParams(request.tokenId, true);
+        SupplyLogic.executeSupplyERC721(
+            ps._reserves,
+            userConfig,
+            DataTypes.ExecuteSupplyERC721Params({
+                asset: request.collection,
+                tokenData: tokenData,
+                onBehalfOf: request.initiator,
+                payer: address(0),
+                referralCode: 0
+            })
+        );
+
+        //update status here to prevent consuming gas for saving requestHash or calculating requestHash twice
+        ps._blurExchangeRequestStatus[requestHash] = DataTypes
+            .BlurBuyWithCreditRequestStatus
+            .Initiated;
+
+        //emit event
+        emit BlurExchangeRequestInitiated(
+            request.initiator,
+            request.paymentToken,
+            request.listingPrice,
+            request.borrowAmount,
+            request.collection,
+            request.tokenId
+        );
+
+        return request.listingPrice + requestFee - request.borrowAmount;
```

Comment:
Can we save some gas if calculate needCashETH  first and sent into validateInitiateBlurExchangeRequest  ?